types/transaction_target.go

```diff
@@ -0,0 +1,256 @@
+package types
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+
+	"github.com/make-software/casper-go-sdk/types/clvalue"
+	"github.com/make-software/casper-go-sdk/types/key"
+)
+
+const (
+	TransactionTargetTypeNative = iota
+	TransactionTargetTypeStored
+	TransactionTargetTypeSession
+)
+
+const (
+	InvocationTargetTagByHash = iota
+	InvocationTargetTagByName
+	InvocationTargetTagByPackageHash
+	InvocationTargetTagByPackageName
+)
+
+type TransactionTarget struct {
+	// The execution target is a native operation (e.g. a transfer).
+	Native *struct{}
+	// The execution target is a stored entity or package.
+	Stored *StoredTarget `json:"Stored"`
+	// The execution target is the included module bytes, i.e. compiled Wasm.
+	Session *SessionTarget `json:"Session"`
+}
+
+func (t *TransactionTarget) Bytes() ([]byte, error) {
+	result := make([]byte, 0, 8)
+
+	if t.Native != nil {
+		result = append(result, TransactionTargetTypeNative)
+	} else if t.Stored != nil {
+		result = append(result, TransactionTargetTypeStored)
+		if byHash := t.Stored.ID.ByHash; byHash != nil {
+			result = append(result, InvocationTargetTagByHash)
+			result = append(result, byHash.Bytes()...)
+		} else if byName := t.Stored.ID.ByName; byName != nil {
+			result = append(result, InvocationTargetTagByName)
+			result = append(result, clvalue.NewCLString(*byName).Bytes()...)
+		} else if byPackageHash := t.Stored.ID.ByPackageHash; byPackageHash != nil {
+			result = append(result, InvocationTargetTagByPackageHash)
+			result = append(result, byPackageHash.Addr.Bytes()...)
+			if byPackageHash.Version != nil {
+				result = append(result, clvalue.NewCLUInt32(*byPackageHash.Version).Bytes()...)
+			}
+		} else if byPackageName := t.Stored.ID.ByPackageName; byPackageName != nil {
+			result = append(result, InvocationTargetTagByPackageName)
+			result = append(result, clvalue.NewCLString(byPackageName.Name).Bytes()...)
+			if byPackageHash.Version != nil {
+				result = append(result, clvalue.NewCLUInt32(*byPackageName.Version).Bytes()...)
+			}
+		}
+		result = append(result, t.Stored.Runtime.RuntimeTag())
+	} else if t.Session != nil {
+		result = append(result, TransactionTargetTypeSession)
+		bytes, err := hex.DecodeString(t.Session.ModuleBytes)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, clvalue.NewCLUInt32(uint32(len(bytes))).Bytes()...)
+		result = append(result, clvalue.NewCLByteArray(bytes).Bytes()...)
```

Comment:
does this work for an empty module_bytes array? (probably doesn't make sense tho have empty module_bytes though)